ui/src/app/views/workspace-share/component.tsx

```diff
@@ -0,0 +1,409 @@
+      workspaceFound: (this.props.workspace !== null),
+      workspaceUpdateConflictError: false,
+      workspace: this.props.workspace,
+      searchTerm: '',
+    };
+    this.searchTermChangedEvent = fp.debounce(300, this.userSearch);
+  }
+
+  save(): void {
+    if (this.state.usersLoading) {
+      return;
+    }
+    this.setState({usersLoading: true, workspaceShareError: false});
+    workspacesApi().shareWorkspace(this.state.workspace.namespace,
+      this.state.workspace.id,
+      {workspaceEtag: this.state.workspace.etag, items: this.state.workspace.userRoles})
+      .then((resp: ShareWorkspaceResponse) => {
+        const updatedWorkspace = this.state.workspace;
+        fp.set('etag', resp.workspaceEtag, updatedWorkspace);
+        fp.set('userRoles', resp.items, updatedWorkspace);
+        this.setState({usersLoading: false, userNotFound: '',
+          searchTerm: '', workspace: updatedWorkspace});
+        this.props.closeFunction();
+      }).catch(error => {
+        if (error.status === 400) {
+          this.setState({userNotFound: error});
+        } else if (error.status === 409) {
+          this.setState({workspaceUpdateConflictError: true});
+        } else {
+          this.setState({workspaceShareError: true});
+        }
+        this.setState({usersLoading: false});
+      });
+  }
+
+  removeCollaborator(user: UserRole): void {
+    if (!this.state.usersLoading) {
+      const position = this.state.userRolesList.findIndex((userRole) => {
+        return user.email === userRole.email;
+      });
+      this.setState(({userRolesList}) => (
+        {userRolesList: fp.pullAt(position, userRolesList)}));
+    }
+  }
+
+  reloadWorkspace(): void {
+    workspacesApi().getWorkspace(this.state.workspace.namespace, this.state.workspace.id)
+      .then((workspaceResponse) => {
+        this.setState({workspace: workspaceResponse.workspace});
+        this.resetModalState();
+      })
+      .catch(error => {
+        if (error.status === 404) {
+          this.setState({workspaceFound: false});
+        }
+      });
+  }
+
+  addCollaborator(user: User): void {
+    const userRole: UserRole = {givenName: user.givenName, familyName: user.familyName,
+      email: user.email, role: WorkspaceAccessLevel.READER};
+    this.setState(({userRolesList}) => (
+      {searchTerm: '', autocompleteLoading: false, autocompleteUsers: [],
+        userRolesList: fp.concat(userRolesList, [userRole])}
+    ));
+  }
+
+  searchTermChangedEvent($event: string) {
+    this.userSearch($event);
+  }
+
+  userSearch(value: string): void {
+    this.setState({autocompleteLoading: true, autocompleteUsers: [], searchTerm: value});
+    if (!value.trim()) {
+      this.setState({autocompleteLoading: false});
+      return;
+    }
+    userApi().user(this.state.searchTerm)
+      .then((response) => {
+        this.setState({autocompleteLoading: false});
+        response.users = fp.differenceWith((a, b) => {
```

Comment:
`differenceWith` is probably not correct here as it's going to pull in elements from the user list as well. I'd throw the `userRolesList` emails into a `Set` then filter against that.

Reply:
I'm a little confused here. Isn't the behaviour we want is that the autocomplete returns users matching search that are NOT already in the userRolesList? and isn't that what differenceWith does? Feel free to DM me if I'm around.